ext/nokogiri/xml_node.c

```diff
@@ -33,6 +43,30 @@ _xml_node_mark(xmlNodePtr node)
   }
 }
 
+#ifdef HAVE_RB_GC_LOCATION
+static void
+_xml_node_update_references(xmlNodePtr node)
+{
+  if (node->_private) {
+    node->_private = (void *)rb_gc_location((VALUE)node->_private);
+  }
+}
+#endif
+
+typedef void (*gc_callback_t)(void *);
+
+static const rb_data_type_t nokogiri_node_type = {
+    "Nokogiri/XMLNode",
+    {(gc_callback_t)_xml_node_mark, (gc_callback_t)_xml_node_dealloc, 0,
+#ifdef HAVE_RB_GC_LOCATION
+        (gc_callback_t)_xml_node_update_references
```

Comment:
I think this code has undefined behavior. Although you can cast between incompatible function pointers, you can't legally call functions through them. C99, Annex J's list of undefined behavior includes
> A pointer is used to call a function whose type is not compatible with the pointed-to type (6.3.2.3).

And indeed §6.3.2.3 paragraph 8 says
> A pointer to a function of one type may be converted to a pointer to a function of another type and back again; the result shall compare equal to the original pointer. If a converted pointer is used to call a function whose type is not compatible with the pointed-to type, the behavior is undefined.

I believe all of `_xml_node_mark`, `_xml_node_dealloc`, and `_xml_node_update_references` need to take a `void *` argument and then assign the argument to an `xmlNodePtr` local.

Reply:
I think the function types are compatible though?  They all take a single pointer as the parameter (IOW the width of a `void *` is the same width as an `xmlNodePtr` so there shouldn't be any problem wrt C calling convention).  I don't have any strong feelings wrt taking a `void *` and casting in the function body, but I don't think this behavior is undefined.

Reply:
The types are not compatible according to C. It's undefined in exactly the same way [this is](https://stackoverflow.com/questions/2561697/warning-when-using-qsort-in-c). The problem is that [function pointer types have to have compatible arguments](https://en.cppreference.com/w/c/language/type#Compatible_types) and `void *` is not compatible with `xmlNodePtr`.

In fact, C doesn't mandate that `void *` and `xmlNodePtr` be the same size; although they probably are on any platform on which Ruby runs.

Reply:
You shouldn't need to cast the `void *` to assign it to another pointer though.

Reply:
Should be fixed in 0fd5321931c80250743d322381eb571a90a7f926!

Reply:
Thanks!